ee/modules/600-flant-integration/images/flant-pricing/hooks/controller_metrics.py

```diff
@@ -0,0 +1,189 @@
+# Licensed under the Deckhouse Platform Enterprise Edition (EE) license.
+# See https://github.com/deckhouse/deckhouse/blob/main/ee/LICENSE
+#
+# This hook is responsible for generating metrics for d8 controllers resource consumption.
+
+
+from dataclasses import dataclass
+from abc import ABC, abstractmethod
+from typing import List, Dict, Any, TypeVar
+
+from shell_operator import hook
+from utils import prometheus_query_value, prometheus_metric_builder, prometheus_function_builder
+
+
+@dataclass
+class Controller:
+    name: str
+    namespace: str
+    kind: str
+    module: str
+
+
+MetricCollectorT = TypeVar("MetricCollectorT", bound="AbstractMetricCollector")
+
+
+class AbstractMetricCollector(ABC):
+
+    metric_group = "group_d8_controller_metrics"
+    cpu_metric_name = "flant_pricing_controller_average_cpu_usage_seconds"
+    memory_metric_name = "flant_pricing_controller_average_memory_working_set_bytes:without_kmem"
+    interval = "5m"
+
+    def collect(self, ctx: hook.Context, controllers: List[Controller]):
+        '''Export metrics to hook context from Controllers list'''
+
+        ctx.metrics.expire(self.metric_group)
+        for ctrl in controllers:
+            # Export metrics
+            ctx.metrics.collect({
+                "name": self.cpu_metric_name,
+                "group": self.metric_group,
+                "set": self.get_cpu_controller_consumption(ctrl),
+                "labels": controller_metric_labels(ctrl),
+            })
+
+            ctx.metrics.collect({
+                "name": self.memory_metric_name,
+                "group": self.metric_group,
+                "set": self.get_memory_controller_consumption(ctrl),
+                "labels": controller_metric_labels(ctrl),
+            })
+
+    @abstractmethod
+    def get_cpu_controller_consumption(self, controller: Controller) -> float:
+        raise NotImplementedError('define get_cpu_controller_consumption to use this base class')
+
+    @abstractmethod
+    def get_memory_controller_consumption(self, controller: Controller) -> float:
+        raise NotImplementedError('define get_memory_controller_consumption to use this base class')
+
+
+class MetricCollector(AbstractMetricCollector):
+    resource_consumption_query = '''
+    sum (
+        ( {} )
+        + on(pod) group_left(controller_name, controller_type)
+        ( {} * 0 )
+    )
+    '''
+
+    def get_cpu_controller_consumption(self, controller: Controller) -> float:
+        '''Query prometheus for controller cpu consumption'''
+
+        return prometheus_query_value(
+            query=self.resource_consumption_query.format(
+                prometheus_function_builder(
+                    func="rate",
+                    metric=resource_metric(
+                        metric_name="container_cpu_usage_seconds_total",
+                        controller=controller,
+                    ),
+                    interval=self.interval,
+                ),
```

Comment:
I think this place is too bloated with wrapping functions. It is hard to get what is going on, what is the resulting expression. I'd make it more explicit.

Reply:
E.g. we always use interval. Why it is optional in `prometheus_function_builder`? I'd write its content right here instead of extracting a function.

Also, `resource_metric` adds almost nothing as for me.

In `query.format(..., ...)`, we have two arguments, but it is a little too noisy to read in 8 lines.

I don't think it is a blocker for merge, but I'd make this place a more clean.